portal/src/main/webapp/js/src/oql/oqlfilter.js

```diff
@@ -0,0 +1,413 @@
+// Requires parser to be window object, window.oql_parser
+// Heavily dependent on OQL PEGjs specification
+
+window.OQL = (function () {
+
+    var parseOQLQuery = function (oql_query, opt_default_oql) {
+	/*	In: - oql_query, a string, an OQL query
+		    - opt_default_oql, a string, default OQL to add to any empty line
+		Out: An array, with each element being a parsed OQL line, with 
+		    all 'DATATYPES' lines applied to subsequent lines and removed.
+	*/
+	opt_default_oql = opt_default_oql || "";
+	var parsed = window.oql_parser.parse(oql_query);
+
+	var datatypes_alterations = false;
+	for (var i = 0; i < parsed.length; i++) {
+	    if (parsed[i].gene.toLowerCase() === "datatypes") {
+		datatypes_alterations = parsed[i].alterations;
+	    } else if (datatypes_alterations && !parsed[i].alterations) {
+		parsed[i].alterations = datatypes_alterations;
+	    }
+	}
+
+	if (opt_default_oql.length > 0) {
+	    for (var i = 0; i < parsed.length; i++) {
+		if (!parsed[i].alterations) {
+		    parsed[i].alterations = window.oql_parser.parse("DUMMYGENE:" + opt_default_oql + ";")[0].alterations;
+		}
+	    }
+	    ;
+	}
+
+	return parsed.filter(function (parsed_line) {
+	    return parsed_line.gene.toLowerCase() !== "datatypes";
+	});
+    };
+    
+    var parsedOQLAlterationToSourceOQL = function(alteration) {
+	if (alteration.alteration_type === "cna") {
+	    return alteration.constr_val;
+	} else if (alteration.alteration_type === "mut") {
+	    return "MUT" + (alteration.constr_rel ? alteration.constr_rel + alteration.constr_val : "");
+	} else if (alteration.alteration_type === "exp") {
+	    return "EXP" + alteration.constr_rel + alteration.constr_val;
+	} else if (alteration.alteration_type === "prot") {
+	    return "PROT" + alteration.constr_rel + alteration.constr_val;
+	}
+    };
+    var unparseOQLQueryLine = function (parsed_oql_query_line) {
+	var ret = "";
+	var gene = parsed_oql_query_line.gene;
+	var alterations = parsed_oql_query_line.alterations;
+	ret += gene;
+	if (alterations.length > 0) {
+	    ret += ": " + alterations.map(parsedOQLAlterationToSourceOQL).join(" ");
+	}
+	ret += ";";
+	return ret;
+    };
+    
+    /* For the methods isDatumWantedByOQL, ..., the accessors argument is as follows:
+	     * accessors = {
```

Comment:
Do you mean the conversion back to source OQL? That's for annotating in the portal with the source OQL, for example if there are multiple instances of the same gene in the query and you want to know which track is which set of alterations.

As far as datum methods, the approach I've taken isn't object-oriented.

Does that answer your question?